services/upgrader/upgrader.go

```diff
@@ -0,0 +1,250 @@
+	if err != nil {
+		return errors.New("error getting the executable path")
+	}
+	fmt.Println(executablePath)
+	executableInfo, err := os.Stat(executablePath)
+	if err != nil {
+		return errors.New("error getting the executable info")
+	}
+	stat, ok := executableInfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return errors.New("error getting the executable info")
+	}
+	fileUID := int(stat.Uid)
+	fileUser, err := user.LookupId(strconv.Itoa(fileUID))
+	if err != nil {
+		return errors.New("error getting the executable info")
+	}
+
+	mattermostUID := os.Getuid()
+	mattermostUser, err := user.LookupId(strconv.Itoa(mattermostUID))
+	if err != nil {
+		return errors.New("error getting the executable info")
+	}
+
+	mode := executableInfo.Mode()
+	if fileUID != mattermostUID && mode&(1<<1) == 0 && mode&(1<<7) == 0 {
+		return NewInvalidPermissions("invalid-user-and-permission", executablePath, mattermostUser.Username, fileUser.Username)
+	}
+
+	if fileUID != mattermostUID && mode&(1<<1) == 0 && mode&(1<<7) != 0 {
+		return NewInvalidPermissions("invalid-user", executablePath, mattermostUser.Username, fileUser.Username)
+	}
+
+	if fileUID == mattermostUID && mode&(1<<7) == 0 {
+		return NewInvalidPermissions("invalid-permission", executablePath, mattermostUser.Username, fileUser.Username)
+	}
+	return nil
+}
+
+func canIUpgrade() error {
+	if runtime.GOARCH != "amd64" {
+		return NewInvalidArch()
+	}
+	if runtime.GOOS != "linux" {
+		return NewInvalidArch()
+	}
+	return canIWriteTheExecutable()
+}
+
+func CanIUpgradeToE0() error {
+	if err := canIUpgrade(); err != nil {
+		mlog.Error("Unable to upgrade from TE to E0", mlog.Err(err))
+		return err
+	}
+	if model.BuildEnterpriseReady == "true" {
+		mlog.Warn("Unable to upgrade from TE to E0 because the server is already in E0")
+		return errors.New("You can't upgrade your code to enterprise because you are already in enterprise code")
+	}
+	return nil
+}
+
+func UpgradeToE0() error {
+	if !atomic.CompareAndSwapInt32(&upgrading, 0, 1) {
+		mlog.Warn("Trying to upgrade while other upgrade is running")
+		return errors.New("One upgrade is already running.")
+	}
+	defer atomic.CompareAndSwapInt32(&upgrading, 1, 0)
+
+	upgradePercentage = 1
+	upgradeError = nil
+
+	executablePath, err := os.Executable()
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = errors.New("error getting the executable path")
+		mlog.Error("Unable to get the mattermost executable path", mlog.Err(err))
+		return err
+	}
+
+	filename, err := download(getCurrentVersionTgzUrl())
+	if err != nil {
+		if filename != "" {
+			os.Remove(filename)
+		}
+		upgradeError = fmt.Errorf("error downloading the new version (percentage: %d)", upgradePercentage)
+		mlog.Error("Unable to download the mattermost server version", mlog.Int64("percentage", upgradePercentage), mlog.String("url", getCurrentVersionTgzUrl()), mlog.Err(err))
+		upgradePercentage = 0
+		return err
+	}
+	defer os.Remove(filename)
+
+	err = extractBinary(executablePath, filename)
```

Comment:
After downloading, how do we ensure this is an archive from Mattermost?  After downloading, we should ensure that two things are correct:

1. The SHA256 hash
2. The GPG signature

Otherwise we are opening ourselves up to any MitM attacks.

Reply:
Good point

Reply:
@DSchalla shouldn't be enough the GPG signature? I mean, if the signature is ok, the binary is generated by us, that verifies 2 things, the data integrity and the author.

Reply:
@jespino Yes, you are absolutely correct since we sign our actual archives. :+1:

Reply:
This is now done.